dds_web/utils.py

```diff
@@ -749,3 +749,61 @@ def use_sto4(unit_object, project_object) -> bool:
 
     flask.current_app.logger.info(f"{project_id_logging}sto2")
     return False
+
+
+def add_uploaded_files_to_db(proj_in_db, log):
+    from dds_web import db
+    from dds_web.api.api_s3_connector import ApiS3Connector
+
+    errors = {}
+    files_added = []
+    for file, vals in log.items():
+        status = vals.get("status")
+        if not status or not status.get("failed_op") == "add_file_db":
+            continue
+
+        with ApiS3Connector(project=proj_in_db) as s3conn:
+            try:
+                _ = s3conn.resource.meta.client.head_object(
+                    Bucket=s3conn.project.bucket, Key=vals["path_remote"]
+                )
+            except botocore.client.ClientError as err:
+                if err.response["Error"]["Code"] == "404":
+                    errors[file] = {"error": "File not found in S3", "traceback": err.__traceback__}
+            else:
+                file_object = models.File.query.filter(
+                    sqlalchemy.and_(
+                        models.File.name == sqlalchemy.func.binary(file),
+                        models.File.project_id == proj_in_db.id,
+                    )
+                ).first()
+                if file_object:
+                    errors[file] = {"error": "File already in database."}
+                else:
+                    new_file = models.File(
+                        name=file,
+                        name_in_bucket=vals["path_remote"],
+                        subpath=vals["subpath"],
+                        project_id=proj_in_db.id,
+                        size_original=vals["size_raw"],
+                        size_stored=vals["size_processed"],
+                        compressed=not vals["compressed"],
+                        public_key=vals["public_key"],
+                        salt=vals["salt"],
+                        checksum=vals["checksum"],
+                    )
+                    new_version = models.Version(
+                        size_stored=new_file.size_stored, time_uploaded=datetime.datetime.utcnow()
+                    )
+                    proj_in_db.file_versions.append(new_version)
+                    proj_in_db.files.append(new_file)
+                    new_file.versions.append(new_version)
+
+                    db.session.add(new_file)
+                    try:
+                        db.session.commit()
+                        files_added.append(new_file)
+                    except (sqlalchemy.exc.IntegrityError, sqlalchemy.exc.OperationalError) as err:
+                        errors[file] = {"error": str(err)}
```

Comment:
Reply to https://github.com/ScilifelabDataCentre/dds_web/pull/1472/files#r1358200649

> this works inside the last 'else' block, not sure why 'db.session.commit()' was outside before, and whether we need 'db.session.rollback()' in 'except'

Good catch. We need this over all the database queries though so we need to move the `try` up like this. And yes, I think we do need the rollback.
```suggestion
                try:
                    file_object = models.File.query.filter(
                        sqlalchemy.and_(
                            models.File.name == sqlalchemy.func.binary(file),
                            models.File.project_id == proj_in_db.id,
                        )
                    ).first()
                    if file_object:
                        errors[file] = {"error": "File already in database."}
                    else:
                        new_file = models.File(
                            name=file,
                            name_in_bucket=vals["path_remote"],
                            subpath=vals["subpath"],
                            project_id=proj_in_db.id,
                            size_original=vals["size_raw"],
                            size_stored=vals["size_processed"],
                            compressed=not vals["compressed"],
                            public_key=vals["public_key"],
                            salt=vals["salt"],
                            checksum=vals["checksum"],
                        )
                        new_version = models.Version(
                            size_stored=new_file.size_stored, time_uploaded=datetime.datetime.utcnow()
                        )
                        proj_in_db.file_versions.append(new_version)
                        proj_in_db.files.append(new_file)
                        new_file.versions.append(new_version)

                        db.session.add(new_file)
                        db.session.commit()
                        files_added.append(new_file)
                except (sqlalchemy.exc.IntegrityError, sqlalchemy.exc.OperationalError) as err:
                    errors[file] = {"error": str(err)}
                    db.session.rollback()
```